# Modelos Lineales Generalizados en Python
# Regresión Gamma

<img src="https://raw.githubusercontent.com/fhernanb/fhernanb.github.io/master/docs/logo_unal_color.png" alt="drawing" width="200"/>

Aquí se muestran varios ejemplos de como usar Python para ajustar un modelo lineal generalizado. 

Las explicaciones mostradas aquí están basadas en un video de YouTube https://www.youtube.com/watch?v=__oC5IRCFKI

Las librerías necesarias son las siguientes:

In [1]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.formula.api import glm

Otras librerías que se usarán en los ejemplos son:

In [2]:
import pandas as pd

## Ejemplo

En esta actividad vamos a utilizar los datos de los cangrejos presentados la sección 4.7 de Agresti (2015). El objetivo es ajustar el siguiente modelo:

\begin{align}
Price_i &\sim Gamma(\mu_i, k), \\ 
\mu_i &= \beta_0 + \beta_1 \, size_i + \beta_2 \, new_i + \beta_3 \, size_i \, new_i
\end{align}

El objetivo es modelar el $Precio$ de una casa en función de algunas de sus características. Abajo una figura ilustrativa.

<img src="casas.jpg" alt="drawing" width="200"/>

Lo primero que usted debe hacer es leer la base de datos.

In [3]:
file = 'http://users.stat.ufl.edu/~aa/glm/data/Houses.dat'
datos = pd.read_csv(file, sep='\s+', header=0)
datos.head()

case  taxes  beds  baths  new  price  size
0     1   3104     4      2    0  279.9  2048
1     2   1173     2      1    0  146.5   912
2     3   3076     4      2    0  237.7  1654
3     4   1608     3      2    0  200.0  2068
4     5   1454     3      3    0  159.9  1477

Para ver el tamaño de la base de datos

In [4]:
datos.shape

(100, 7)

Para ajustar el modelo:

In [5]:
mod1 = smf.glm(formula='price ~ size + new + size*new', data=datos, 
               family=sm.families.Gamma(link=sm.families.links.identity()))
mod1 = mod1.fit()
mod1.summary()

C:\Users\fhern\Anaconda3\lib\site-packages\statsmodels\genmod\generalized_linear_model.py:296: DomainWarning: The identity link function does not respect the domain of the Gamma family.
  DomainWarning)


<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:                  price   No. Observations:                  100
Model:                            GLM   Df Residuals:                       96
Model Family:                   Gamma   Df Model:                            3
Link Function:               identity   Scale:                         0.11020
Method:                          IRLS   Log-Likelihood:                -519.05
Date:                Thu, 29 Apr 2021   Deviance:                       10.563
Time:                        20:09:39   Pearson chi2:                     10.6
No. Iterations:                    11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -7.4509     12.974     -0.574      0.566     -32.880      17.978
size           0.0945      0.010      9.396      0.000       0.075       0.114
new          -77.9046     64.582     -1.206      0.228    -204.483      48.674
size:new       0.0649      0.037      1.769      0.077      -0.007       0.137
==============================================================================
"""

Usando los resultados de la tabla anterior podemos escribir el modelo

\begin{align}
Precio_i &\sim Gamma(\hat{\mu}_i, \hat{k}), \\ 
\hat{\mu}_i &= -7.4509 + 0.0945 \, size_i -77.9046 \, new_i + 0.0649 \, size_i \, new_i \\
\hat{k} &= 1/0.11020 = 9.07441
\end{align}

## Funciones de enlace disponibles

Para conocer otras posibles funciones de enlace se puede utilizar la siguiente instrucción:

In [6]:
sm.families.family.Gamma.links

[statsmodels.genmod.families.links.log,
 statsmodels.genmod.families.links.identity,
 statsmodels.genmod.families.links.inverse_power]